kernel/pte_flags/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+//! This crate defines the structure of page table entry (PTE) flags on x86_64 and aarch64.
+//! 
+//! This crate offers two main types:
+//! * [`PteFlags`]: the set of bit flags that apply to all architectures.
+//! * [`PteFlagsX86_64`] or [`PteFlagsAarch64`]: the arch-specific set of bit flags
+//!   that apply to only the given platform.
+//! 
+//! ## Type conversions
+//! *Notably*, you can convert to and from these architecture-specific types
+//! and architecture-generic type easily.
+//! [`PteFlags`] can be losslessly converted into [`PteFlagsX86_64`] or [`PteFlagsAarch64`],
+//! with the typical [`From`] and [`Into`] traits.
+//! This makes it possible to set general architecture-indepedent flags first,
+//! and then convert it in order to set more architecture-specific flags.
+//! 
+//! You can also convert [`PteFlagsX86_64`] or [`PteFlagsAarch64`] into [`PteFlags`],
+//! but it may be lossy as only the bit flags defined in [`PteFlags`] are preserved.
+//! 
+//! ## aarch64 considerations
+//! When converting from [`PteFlags`] to [`PteFlagsAarch64`],
+//! certain bits will be set by default;
+//! see [`PteFlagsAarch64::from()`] for more information.
+//! 
+//! For aarch64, the definition of these flags assumes that the [MAIR] index 0
+//! has a "DEVICE nGnRE" entry, and [MAIR] Index 1 has a Normal + Outer Shareable entry.
+//!
+//! [MAIR]: https://docs.rs/cortex-a/latest/cortex_a/registers/MAIR_EL1/index.html
+
+#![no_std]
+#![feature(doc_cfg)]
+
+use cfg_if::cfg_if;
+use bitflags::bitflags;
+
+cfg_if!{ if #[cfg(any(target_arch = "x86_64", doc))] {
+    mod pte_flags_x86_64;
+    pub use pte_flags_x86_64::PteFlagsX86_64;
+}}
+cfg_if!{ if #[cfg(any(target_arch = "aarch64", doc))] {
+    mod pte_flags_aarch64;
+    pub use pte_flags_aarch64::PteFlagsAarch64;
+}}
+
+#[cfg(target_arch = "x86_64")]
+use pte_flags_x86_64::PteFlagsX86_64 as PteFlagsArch;
+#[cfg(target_arch = "aarch64")]
+use pte_flags_aarch64::PteFlagsAarch64 as PteFlagsArch;
```

Comment:
Could these be re-exported publicly? So that I can do:
```rust
impl PageTableEntry {
    pub fn flags(&self) -> PteFlags {
        PteFlagsArch::from_bits_truncate(self.0).into()
    }
}
```
in `page_table_entry`

Reply:
hmm so i considered exporting this publicly but couldn't think of a good use case for it. My primary concern was that users would be tempted to use `PteFlagsArch` everywhere when they really should be using `PteFlagsX86_64` or `PteFlagsAarch64` explicitly, which would then lead to really confusing compilation errors on a different platform than the one they initially were testing with, because the definition of `PteFlagsArch` would change depending on target platform. Generally that's indicative of a bad design.

Your use case here is definitely valid, but I think that you'll need a fair bit of aarch64-specific code in the `page_table_entry` crate. For example: several things have to change:
* `frame_value()` needs a different impl on aarch64 than x86 because more bits need to be masked out
* `set_entry()` needs to change (depending on how you implement it), e.g., you may choose to set the `PAGE_DESCRIPTOR` bit there (or elsewhere, higher up the call stack)
* `set_flags()` will be different too; we actually need to mask all of the bits in the arch-specific `PteFlags` type in order to make sure they're correctly set. The existing code only works for x86

Thus, since you'll need tons of arch-specific cfg, I think it's best to keep things explicit, which is also more Rust-y in nature. As such, we'll leave `PteFlagsArch` typedef private.